### PR TITLE
OpenTelemetry proto formatter for access logging

### PR DESCRIPTION
**Pull request by itamarkam**
Signed-off-by: Itamar Kaminski <itamark@google.com>

Commit Message: A dedicated OpenTelemetryFormatter used for access logging
Additional Description: Before this change, the OpenTelemetry access logger used the "StructFormatter", so it had to convert OT format to Struct and back which wasted a lot of CPU. This change copies most of the logic from the StructFormatter and creates a formatter that uses OpenTelemetry e2e.
Risk Level: Low - only used in OT access logging, without I/O changes.
Testing:
Unit tests and integration tests
Microbenchmarks - A new microbenchamrks for the OpenTelemetry formatter was added. In order to test the difference in the access logger performance (that doesn't have a benchmark), I created an additional microbenchmark (locally) that performs the OpenTelemetry access logger's flow from before this change - Take OTel config, convert to Struct, format using StructFormatter, convert the output back to OTel.
Results show a significant difference, reducing time and CPU by ~30x!:
$ bazel run -c opt test/common/formatter:substitution_formatter_speed_test  --config=remote

Benchmark                                   Time             CPU   Iterations
--------------------------------------- --------------------------------------
BM_OldOpenTelemetryAccessLogFormatter            182017 ns       181623 ns         5446
BM_OpenTelemetryAccessLogFormatter                      6001 ns          6000 ns        98773

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/formatter/substitution_formatter.h [hunk 1]
```diff
@@ -241,6 +243,95 @@ class StructFormatter {
...
+  using OpenTelemetryFormatValue =
+      absl::variant<const std::vector<FormatterProviderPtr>, const OpenTelemetryFormatMapWrapper,
+                    const OpenTelemetryFormatListWrapper>;
+  // Although not required for Struct/JSON, it is nice to have the order of
+  // properties preserved between the format and the log entry, thus std::map.
+  using OpenTelemetryFormatMap = std::map<std::string, OpenTelemetryFormatValue>;
+  using OpenTelemetryFormatMapPtr = std::unique_ptr<OpenTelemetryFormatMap>;
+  struct OpenTelemetryFormatMapWrapper {
+    OpenTelemetryFormatMapPtr value_;
+  };
```
Comment (rojkov) [inline source/common/formatter/substitution_formatter.h:282]: A drive-be question: what's the benefit of `OpenTelemetryFormatMapWrapper` over `OpenTelemetryFormatMapPtr`?

Comment (itamarkam) [reply source/common/formatter/substitution_formatter.h:282]: I just copied the whole StructFormatter as is and converted the types to match OpenTelemetry logging.
I tried templating it but it adds a lot of boiler plate which I'm not sure is worth the trouble.

**Comment by itamarkam** [comment]
> drive-by: do you have a microbenchmark showing the improvement from this approach? It looks like you do. Can you put results into the PR description?

With the new OTel formatter, I added a microbenchmark for the formatter. The benchmark is for the formatter, not the access logger, so there's nothing to compare to, so I added a microbenchmark locally that does what the access logger used to do beforehand - take OT config, convert to Struct, format with StructFormatter, convert output back to OT.
Results show a significant difference, reducing time and CPU by ~30x! (also added to the description):
$ bazel run -c opt test/common/formatter:substitution_formatter_speed_test  --config=remote

-----------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations
-----------------------------------------------------------------------------
BM_OldOpenTelemetryAccessLogFormatter            182017 ns       181623 ns         5446
BM_OpenTelemetryAccessLogFormatter                      6001 ns          6000 ns        98773

**Review by mattklein123 — DISMISSED** [review]
Thanks LGTM. Should this change have a release note?

/wait

**Comment by itamarkam** [comment]
> @itamarkam please never force push as it makes reviews much more difficult. Thanks!

Yeah you're right, sorry for that. There were no code changes requested so I thought it wouldn't matter that much but it does make it more complicated.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**